proposals/state-machine-restriction-updates.md

```diff
@@ -0,0 +1,167 @@
+[summary]: #summary
+
+- Allow `ref`/`ref struct` locals and `unsafe` blocks in iterators and async methods
+  provided they are used in code segments without any `yield` or `await`.
+- Warn about `yield` inside `lock`.
+- Consolidate spec: disallow pointers to hoisted variables
+  (pointers to captured variables are already disallowed).
+
+## Motivation
+[motivation]: #motivation
+
+It is not necessary to disallow `ref`/`ref struct` locals and `unsafe` blocks in async/iterator methods
+if they are not used across `yield` or `await`, because they do not need to be hoisted.
+
+```cs
+async void M()
+{
+    await ...;
+    ref int x = ...; // error previously, proposed to be allowed
+    x.ToString();
+    await ...;
+    // x.ToString(); // still error
+}
+```
+
+On the other hand, having `yield` inside a `lock` means the caller also holds the lock while iterating which might lead to unexpected behavior.
+This is even more problematic in async iterators where the caller can `await` between iterations, but `await` is not allowed in `lock`.
+See also https://github.com/dotnet/roslyn/issues/72443.
+
+```cs
+lock (this)
+{
+    yield return 1; // warning proposed
+}
+```
+
+## Detailed design
+[design]: #detailed-design
+
+[§13.3.1 Blocks > General][blocks-general]:
+
+> ~~It is a compile-time error for an iterator block to contain an unsafe context ([§23.2][unsafe-contexts]).~~
+> An iterator block always defines a safe context, even when its declaration is nested in an unsafe context.
+
+[§13.6.2.4 Ref local variable declarations][ref-local]:
+
+> ~~It is a compile-time error to declare a ref local variable, or a variable of a `ref struct` type,
+> within a method declared with the *method_modifier* `async`, or within an iterator ([§15.14][iterators]).~~
+> **It is a compile-time error to declare and use a ref local variable, or a variable of a `ref struct` type
+> across `await` or `yield` statements.**
+
+[§13.13 The lock statement][lock-statement]:
+
+> [...]
+> 
+> **A warning is reported (as part of the next warning wave) when a `yield` statement
+> ([§13.15][yield-statement]) is used inside the body of a `lock` statement.**
+
+No change in the spec is needed to allow `unsafe` blocks which do not contain `await`s in async methods,
+because the spec has never disallowed `unsafe` blocks in async methods.
+However, the spec should have always disallowed `await` inside `unsafe` blocks
+(it had already disallowed `yield` in `unsafe` in [§13.3.1][blocks-general] as cited above), for example:
+
+[§15.15.1 Async Functions > General][async-funcs-general]:
+
+> It is a compile-time error for the formal parameter list of an async function to specify
+> any `in`, `out`, or `ref` parameters, or any parameter of a `ref struct` type.
+>
+> **It is a compile-time error for an unsafe context ([§23.2][unsafe-contexts]) to contain `await` or `yield`.**
+
+Furthermore, variables inside async or iterator methods should not be "fixed" but rather "moveable"
```

Comment:
Is this a breaking change? Is this going to require a user to use `fixed` depending on whether the local is hoisted or not?

Reply:
Yes, it is a breaking change. No, the intention is to disallow taking pointers of such locals entirely - that's consistent with how variables captured in lambdas and local functions work today. Although allowing to use `fixed` with those variables (both captured and hoisted) is a possible alternative or future work. I guess there are also other options:
- Require `fixed` for all locals in async methods (no need to distinguish hoisted or not).
- Leave it as is, maybe emit only a warning.

Reply:
I discovered there is already a warning for this in C# 12 warning wave (which is likely why it's not visible in sharplab): https://github.com/dotnet/roslyn/pull/66915

So if we already decided to go with a warning here, I will remove this part of the feature, only leaving a note in alternatives.